### PR TITLE
Rod sphere contact

### PR DESCRIPTION
**Pull request by Rahul-JOON**
This PR attempts to add Rod SPhere contact functionality as per #266.

### CODE REVIEW

**Comment by Rahul-JOON** [comment]
Changes in examples/Binder have been repeated from #294 since this formatting has not been done in the update-0.3.2 branch. These changes are done by auto-formatting while committing.

File: elastica/contact_forces.py [hunk 1]
```diff
@@ -725,3 +863,121 @@ def apply_contact(
...
+        self.friction_coefficient = friction_coefficient
+
+    def _check_systems_validity(
+        self,
+        system_one: SystemType,
+        system_two: AllowedContactType,
+    ) -> None:
+        """
+        This checks the contact order and type of a SystemType object and an AllowedContactType object.
+        For the RodCylinderContact class first_system should be a rod and second_system should be a sphere.
```
Comment (Ali-7800) [inline elastica/contact_forces.py:922]:
```suggestion
        For the RodSphereContact class first_system should be a rod and second_system should be a sphere.
```

File: elastica/contact_forces.py [hunk 2]
```diff
@@ -423,6 +424,143 @@ def _calculate_contact_forces_self_rod(
...
+    friction_coefficient,
+) -> None:
+    # We already pass in only the first n_elem x
+    n_points = x_collection_rod.shape[1]
+    sphere_total_contact_forces = np.zeros((3))
+    sphere_total_contact_torques = np.zeros((3))
+    for i in range(n_points):
+        # Element-wise bounding box
+        x_selected = x_collection_rod[..., i]
+        # x_cylinder is already a (,) array from outised
```
Comment (Ali-7800) [inline elastica/contact_forces.py:454]: update comment

File: tests/test_contact_utils.py [hunk 3]
```diff
@@ -273,10 +274,37 @@ def mock_cylinder_init(self):
...
+
+    "Initializing Sphere"
+
+    """
+    This is a rigid body sphere;,
+    Initial Parameters:
+    radius = 1,
+    center positioned at origin i.e (0, 0, 0),
+    sphere's upright in x,y,z plane thus the director array,
+    stationary cylinder i.e velocity vector is (0, 0, 0),
```
Comment (Ali-7800) [inline tests/test_contact_utils.py:287]: update comment

File: tests/test_contact_functions.py [hunk 4]
```diff
@@ -63,10 +64,37 @@ def mock_cylinder_init(self):
...
+
+    "Initializing Sphere"
+
+    """
+    This is a rigid body sphere;,
+    Initial Parameters:
+    radius = 1,
+    center positioned at origin i.e (0, 0, 0),
+    sphere's upright in x,y,z plane thus the director array,
+    stationary cylinder i.e velocity vector is (0, 0, 0),
```
Comment (Ali-7800) [inline tests/test_contact_functions.py:77]: update comment

**Comment by Ali-7800** [comment]
@Rahul-JOON Can you add tests for the RodSphereContact class? Otherwise looks good to me just some minor comments

File: elastica/contact_forces.py [hunk 5]
```diff
@@ -725,3 +861,120 @@ def apply_contact(
...
+            + system_one.position_collection[..., :-1]
+        )
+        _calculate_contact_forces_rod_sphere(
+            rod_element_position,
+            system_one.lengths * system_one.tangents,
+            system_two.position_collection[..., 0],
+            x_sph,
+            system_two.radius * system_two.director_collection[2, :, 0],
+            system_one.radius + system_two.radius,
+            system_one.lengths + system_two.lengths,
```
Comment (armantekinalp) [inline elastica/contact_forces.py:968]: Length doesn't make much sense of sphere, and it is set as length = 2*radius. Maybe here we replace with following 
```suggestion
            system_one.lengths + 2 * system_two.radius,
```

**Review by armantekinalp — CHANGES_REQUESTED** [review]
minor comments.

File: elastica/contact_forces.py (hunk 5 above)
Comment (Ali-7800) [reply elastica/contact_forces.py:968]: I agree, maybe we should remove it later too cause it only makes sense for cylinders.

File: elastica/contact_forces.py [hunk 6]
```diff
@@ -943,15 +943,15 @@ def apply_contact(self, system_one: RodType, system_two: AllowedContactType):
             system_one.position_collection,
             system_one.radius,
             system_one.lengths,
-            system_two.position_collection,
-            system_two.director_collection,
-            system_two.radius[0],
+            system_two.position,
```
Comment (Ali-7800) [inline elastica/contact_forces.py:946]: It was correct before why did you change it?

**Review by Ali-7800 — CHANGES_REQUESTED** [review]
